Add additional options to installBioc.r

pmoris opened the pull request:
This PR adds additional options to the installBioc.r example script based on the approach used in install2.r. 

The main goal was to be able to use the installBioc.r script as the main method to install bioconductor packages in Dockerfiles, because the standard method that is recommended by Bioconductor (`RUN R -e 'BiocManager::install("scAlign")'`, https://www.bioconductor.org/help/docker/) does not throw an error that causes the docker build process to stop. 

This can result in seemingly successfully built docker images that lack some of the packages which the creator expected to be installed. Since building docker images with r packages is so error prone (due to the need for system libraries that often just requires trial and error), this new script can make the task a lot easier. Builds will fail clearly, rather than having to scroll through the wall of text install log output or trying to manually load each package when running a container.

The script now throws errors when a package fails to install due to missing dependencies, misspelled names, etc. Already installed packages can be skipped. Suggested dependencies can be toggled. Parallel install can be enabled. Download method can be set. 

A few options were added for BiocManager::install()'s unique arguments. A toggle to update other packages that are
outdated. Additional repositories can be prepended to the standard BioC repository (only works for bioc sub-repositories
like software/experiment, see https://rdrr.io/cran/BiocManager/man/install.html). Already installed packages can be forced to redownload and install.

In addition, the library path is now set in the same manner as is done in install2.r (by reading the first entry of .libPaths() and passing it to the install function, rather than setting it to `/usr/local/lib/R/site-library` via `.libPaths()`).

This addresses some of the issues I faced myself (https://github.com/Bioconductor/bioconductor_docker/issues/38, https://github.com/Bioconductor/BiocManager/issues/124, https://github.com/eddelbuettel/littler/issues/93, https://github.com/rocker-org/rocker-versioned2/issues/292) as well as others (https://github.com/rocker-org/rocker-versioned2/issues/370).

I believe my modifications to the script could still benefit from a critical review. Especially the errors that the script now attempts to catch is somewhat guesswork. Based on my own testing, these messages are the most common ones, but perhaps I've missed some others.

eddelbuettel commented:
Thanks for taking the time work on improvements, and for writing such a detail issue and PR.  I will try to take a look.  As you can image I also had my issues with BioC installations but as I am not _primarily_ a BioC user I tended to get by.  I am not sure I myself have a full use case for testing here (_i.e._ 'fail and reinstall' logic which looks promising) but maybe that does not need to stand in the way of a merge as other can test.

Lastly, as it is related and reflects some of my more recent work:  have you seen [r2u](https://eddelbuettel.github.io/r2u/) ? _Full_ and _complete_ binaries with proper dependencies.  I aim for CRAN, have full coverage and because of the overlap also 200+ BioC packages.  Might be possible to grow the BioC side.  You cannot beat `apt` for speed _and_ ease _and_ reliability.

pmoris commented:
Thanks for having a look and for sharing your r2u project. It sounds interesting so I'll definitely explore it in more detail in the future. 

As a side note: I'll be unavailable for a few days, so my future responses will be delayed a bit ;)

eddelbuettel commented on inst/examples/installBioc.r:
```diff
@@ -24,20 +34,93 @@ package 'BiocManger' which has be installed.
 
 if (!requireNamespace("BiocManager", quietly=TRUE)) {
-    stop("Please install 'BiocManager' first, for example via 'install.r BiocManager'.", call.=FALSE)
+    stop("Please install 'BiocManager' first, for example via 'install2.r BiocManager'.", call.=FALSE)
```
Nit:  Why the change on that line?  `install.r BiocManager` also works.

(Minutes later:  Ah well if this script models itself after `install2,r`, maybe?  Oh well -- both will get BiocManager installed...)

eddelbuettel replied:
I revert it back to `install.r` simply because that one is older + simpler + (maybe?) more widely know.  But it is a non-issue.

eddelbuettel commented on inst/examples/installBioc.r:
```diff
@@ -24,20 +34,93 @@ package 'BiocManger' which has be installed.
 Examples:
   installBioc.r -l /tmp/lib S4Vectors               # install into given library
+  installBioc.r --update Biobase                    # install package and update older packages
+  installBioc.r --deps NA --error --skipinstalled   # install package without suggested dependencies, throw an error on installation failure and skip packages that are already present
```
`docopt` prints this line on help so maybe doing 

```
   installBioc.r --deps NA --error --skipinstalled   # install package without suggested dependencies, throw an error
                                                                           # on installation failure and skip packages that are already present
```

(with possible whitespace / space adjustment) is nicer?

eddelbuettel replied:
I am adjusting this now.

eddelbuettel reviewed (changes requested):
This looks great, micro-nit comments notwithstanding.

I take it you tested it?

Can you maybe add a block with a one-liner in this file to ChangeLog?  You will see the format, there are two spaces between the three fields date + name + email.

eddelbuettel reviewed (approved):
I took care of the little nits.  Thanks again for the solid contribution.

eddelbuettel commented:
And I took this as an opportunity to roll up a new release for CRAN which I just sent there, and which should hopefully get processed without any issues.

pmoris commented:
Thanks for the kind words and taking care of the small modifications yourself!

Regarding testing: yes, but mostly ad-hoc. I've been using an earlier draft of this script for a while now without issues. For this PR I added a few of the missing options to make it more on-par with `install2.r`, and I did test out each of the options I added manually. That being said, it is of course possible that I've overlooked something and certain scenarios/ways of using the script will break it....

eddelbuettel commented:
It's all good.  And @eitsupi noticed r-devel wants us to change an error message pattern in the grep expression which he updated in the `install2.r` you worked off, so I also carried that small change over here.  In master, may be a moment before we do a new CRAN release.
